src/libexpr/primops.cc

```diff
@@ -4074,54 +4093,210 @@ void EvalState::createBaseEnv()
+
+          Example:
 
+          ```console
+          $ nix repl
+          Welcome to Nix 2.15.1 Type :? for help.
+
+          nix-repl> builtins.currentTime
+          1683705525
+
+          nix-repl> builtins.currentTime
+          1683705525
+          ```
+
+          The [store path](@docroot@/glossary.md#gloss-store-path) of a derivation depending on `currentTime` will differ for each evaluation.
+        )",
+        .impureOnly = true,
+    });
+
+    if (!evalSettings.pureEval) {
         v.mkString(settings.thisSystem.get());
-        addConstant("__currentSystem", v);
     }
+    addConstant("__currentSystem", v, {
+        .type = nString,
+        .doc = R"(
+          A string in the form `<cpu-architecture>-<operating-systems>` identifying the "current" system.
+          The current system is by default the system where the Nix expression is being evaluated.
+
+          Example of possible `builtins.currentSystem` values:
+
+          - `"i686-linux"`
+          - `"x86_64-darwin"`.
+
+          This builtin constant exists so a Nix expression can be implicitly parameterized what system for which derivations produced by evaluation should be built.
+          If all derivations produced by expression have their system field set with ``builtins.currentSystem` like this:
+
+          ```nix
+          builtins.derivationStrict {
+             ...
+             system = builtins.currentSystem;
+          }
+          ```
+
+          Then, by default, those derivations can be built on the same platform where the expression that produced them was evaluated.
+          This pattern makes this case of evaluating and building on the same machine more ergonomic than e.g. having to pass an argument for the current system on the command-line.
+
+          Pure evaluation eschews this because implicit parameterization like this is impure.
+          The problem of making evaluating and building on the current machine ergonomic can be solved in other ways.
+
+          When available, the current system is defined as the value of the [system](@docroot@/command-ref/conf-file.md#conf-pure-eval) configuration option.
+          That configuration option can be overridden in order to evaluate derivations for different system than the current one instead:
+
+          ```console
+          $ nix-instantiate --option system "i686-darwin" --eval --expr 'builtins.currentSystem'
+          "i686-darwin"
+          ```
+        )",
+        .impureOnly = true,
+    });
 
     v.mkString(nixVersion);
-    addConstant("__nixVersion", v);
+    addConstant("__nixVersion", v, {
+        .type = nString,
+        .doc = R"(
+          The version of Nix.
+
+          If your CLI says, say:
+          ```shell-session
+          $ nix --version
+          nix (Nix) 2.16.0
+          ```
+          Then this would be:
+          ```nix-repl
+          nix-repl> builtins.nixVersion
+          "2.16.0"
```

Comment:
```suggestion      
          For example, where the command line returns the current Nix version,
          
          ```shell-session
          $ nix --version
          nix (Nix) 2.16.0
          ```
          
          the Nix language evaluator returns the same value:
          
          ```nix-repl
          nix-repl> builtins.nixVersion
          "2.16.0"
```